### PR TITLE
Add base class for elliptic boundary conditions

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Opening this PR in draft mode to give context for these PRs:

- [x] #2725
- [x] #2733

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Elliptic/BoundaryConditions/Test_BoundaryCondition.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    return std::make_unique<TestBoundaryCondition>(*this);
+  }
+
+  using argument_tags =
+      tmpl::list<ArgumentTag, VolumeArgumentTag, NonlinearArgumentTag>;
+  using volume_tags = tmpl::list<VolumeArgumentTag>;
+
+  /// [example_poisson_fields]
+  static void apply(
+      const int arg0, const bool arg1, const int arg2,
```
Comment (nilsdeppe) [inline tests/Unit/Elliptic/BoundaryConditions/Test_BoundaryCondition.cpp]: maybe name these `argument_on_face`, `argument_from_volume`, `nonlinear_argument_on_face` or something like that. Since this is used in a snippet I think it would be good to make the names a bit clearer

**Review by nilsdeppe — DISMISSED** [review]
One optional naming clarification, squash if you decide to do it

**Comment by nilsvu** [comment]
Done, thanks @nilsdeppe.

@kidder could you do the second review?

**Comment by nilsvu** [comment]
@nilsdeppe I pushed a fixup that moves some of the code for applying boundary conditions from the test to a function in `src/Elliptic`. I'll use it in the DG operator to start with, but keep it separate so it's easier to replace the way arguments are retrieved from the DataBox later.

File: src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace elliptic {
+/*!
+ * \brief Apply the `boundary_condition` to the `fields_and_fluxes` with
+ * arguments from interface tags in the DataBox.
+ *
+ * This functions assumes the arguments for the `boundary_condition` are stored
+ * in the DataBox in tags
+ * `domain::Tags::Interface<domain::Tags::BoundaryDirectionsInterior<Dim>,
+ * Tag>`. This may turn out not to be the most performant setup, so code that
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp]: `performant` -> `efficient`. I don't know what `performant` means... Chandler Carruth gives a nice rant in his 2014 or 2015 cppcon talk about data structures and algorithms that "performant" doesn't really mean anything :)

File: src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+void apply_boundary_condition(
+    const elliptic::BoundaryConditions::BoundaryCondition<Dim, Registrars>&
+        boundary_condition,
+    const db::DataBox<DbTagsList>& box, const Direction<Dim>& direction,
+    const gsl::not_null<FieldsAndFluxes*>... fields_and_fluxes) noexcept {
+  call_with_dynamic_type<
+      void, typename elliptic::BoundaryConditions::BoundaryCondition<
+                Dim, Registrars>::creatable_classes>(
+      &boundary_condition,
+      [&direction, &box, &fields_and_fluxes...](auto* const derived) noexcept {
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp]: shouldn't the boundary condition be `const` too, not just the pointer? `const auto*const`?

File: src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ * Tag>`. This may turn out not to be the most performant setup, so code that
+ * uses the boundary conditions doesn't have to use this function but can
+ * procure the arguments differently. For example, future optimizations may
+ * involve storing a subset of arguments that don't change during an elliptic
+ * solve in direction-maps in the DataBox, and slicing other arguments to the
+ * interface every time the boundary conditions are applied.
+ */
+template <bool Linearized, size_t Dim, typename Registrars, typename DbTagsList,
+          typename... FieldsAndFluxes>
+void apply_boundary_condition(
```
Comment (nilsdeppe) [inline src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp:36]: this looks like a reasonable way to share the code that handles all the generic type manipulation :)

**Review by nilsdeppe — COMMENTED** [review]
Please go ahead and squash

**Comment by nilsvu** [comment]
Rebased and squashed, thanks @nilsdeppe. Could you take a look at this @kidder?

**Review by kidder — APPROVED** [review]

**Comment by nilsvu** [comment]
@nilsdeppe could you take another look at this?

**Review by nilsdeppe — APPROVED** [review]
